mlrun/runtimes/pod.py

```diff
@@ -1201,6 +1181,91 @@ def with_requests(
+
+        :param tolerations: The user-provided list of tolerations.
+        """
+        preemptible_tolerations = [
+            k8s_client.V1Toleration(
+                key=toleration.get("key"),
+                value=toleration.get("value"),
+                effect=toleration.get("effect"),
+            )
+            for toleration in mlconf.get_preemptible_tolerations()
+        ]
+
+        transform_attribute_to_k8s_class_instance("tolerations", tolerations)
+
+        for toleration in tolerations:
+            if toleration in preemptible_tolerations:
+                self.raise_preemptible_warning(
+                    message=f"Toleration '{toleration.key}: {toleration.value}' "
+                    f"(effect: '{toleration.effect}') may be removed at runtime"
+                )
+
+    def detect_preemptible_affinity(self, affinity: k8s_client.V1Affinity) -> None:
+        """
+        Checks if any provided affinity rules match preemptible affinity configurations.
+        Issues a warning if an affinity rule may be pruned at runtime depending on preemption mode.
+
+        :param affinity: The user-provided affinity object.
+        """
+        preemptible_affinity_terms = generate_preemptible_nodes_affinity_terms()
+
+        if (
+            affinity
+            and affinity.node_affinity
+            and affinity.node_affinity.required_during_scheduling_ignored_during_execution
+        ):
+            user_terms = affinity.node_affinity.required_during_scheduling_ignored_during_execution.node_selector_terms
+
+            for user_term in user_terms:
+                user_expressions = {
+                    (expr.key, expr.operator, tuple(expr.values or []))
+                    for expr in user_term.match_expressions or []
+                }
+
+                for preemptible_term in preemptible_affinity_terms:
+                    preemptible_expressions = {
+                        (expr.key, expr.operator, tuple(expr.values or []))
+                        for expr in preemptible_term.match_expressions or []
+                    }
+
+                    # Ensure operators match and preemptible expressions are present
+                    if user_expressions & preemptible_expressions:
+                        self.raise_preemptible_warning(
+                            message="The selected node affinity constraints may be adjusted at runtime"
+                        )
+
+    def raise_preemptible_warning(self, message: str) -> None:
+        warnings.warn(
+            f"Warning: {message} based on the preemptible node settings configured in your MLRun configuration. "
+            f"This adjustment depends on the function's preemption mode. "
+            f"The list of potential adjusted preemptible selectors can be viewed here: "
+            f"mlrun.mlconf.get_preemptible_node_selector() and mlrun.mlconf.get_preemptible_tolerations()."
+            f"For more information about preemptible mode, see:  "
```

Comment:
If I use `with_node_selection` with mutliple values (e.g. setting NS, affinity and tolerations), I have a possibility to get this message many many times - once for each conflicted label, for each conflicted toleration, and for the affinity.

While this is probably an edge case, how about instead of printing the same warning every time, you accumulate a list of the conflicts, and warn it once (or at least, once for each type)?